File: services/proxy-gateway/admin-panel/src/utils/operatorClient.js
const http = require('http');
const https = require('https');

const BASE_URL = process.env.PROXY_OPERATOR_URL || 'http://127.0.0.1:8091';
const TOKEN = process.env.PROXY_OPERATOR_TOKEN || 'change-me';

function requestJson(method, pathname, body = null, timeout = 60000) {
    return new Promise((resolve, reject) => {
        const url = new URL(pathname, BASE_URL);
        const transport = url.protocol === 'https:' ? https : http;
        const payload = body ? JSON.stringify(body) : null;

        const req = transport.request(
            url,
            {
                method,
                timeout,
                headers: {
                    'Content-Type': 'application/json',
                    'X-Operator-Token': TOKEN,
                    ...(payload ? { 'Content-Length': Buffer.byteLength(payload) } : {}),
                },
            },
            (res) => {
                let raw = '';
                res.on('data', (chunk) => {
                    raw += chunk;
                });
                res.on('end', () => {
                    try {
                        const data = raw ? JSON.parse(raw) : {};
                        if (res.statusCode >= 400) {
                            reject(new Error(data.error || data.detail || `Operator HTTP ${res.statusCode}`));
                            return;
                        }
                        resolve(data);
                    } catch (error) {
                        reject(new Error(`Invalid operator response: ${raw}`));
                    }
                });
            }
        );

        req.on('timeout', () => {
            req.destroy(new Error('Operator timeout'));
        });
        req.on('error', reject);

        if (payload) {
            req.write(payload);
        }
        req.end();
    });
}

class OperatorClient {
    static applyState(proxies) {
        return requestJson('POST', '/api/runtime/apply', { proxies });
    }

    static getHealth() {
        return requestJson('GET', '/health', null, 10000);
    }

    static ensureWildcardCertificate() {
        return requestJson('POST', '/api/runtime/wildcard/ensure', {});
    }

    static restart(name) {
        return requestJson('POST', `/api/runtime/restart/${encodeURIComponent(name)}`, {});
    }
}

module.exports = OperatorClient;

File: apps/platform-control/app/operator_client.py
"""Client for proxy-operator runtime apply jobs."""
from __future__ import annotations

from typing import Any

import httpx

from app.config import get_settings


def _response_error_message(response: httpx.Response) -> str:
    try:
        payload = response.json()
    except ValueError:
        payload = None

    if isinstance(payload, dict):
        detail = payload.get("error") or payload.get("detail") or payload.get("message")
        if detail:
            return str(detail)

    body = (response.text or "").strip()
    if body:
        return body
    return f"{response.status_code} {response.reason_phrase}"


async def apply_proxy_state(proxies: list[dict[str, Any]]) -> dict[str, Any]:
    settings = get_settings()
    async with httpx.AsyncClient(timeout=60.0) as client:
        response = await client.post(
            f"{settings.proxy_operator_url.rstrip('/')}/api/runtime/apply",
            json={"proxies": proxies},
            headers={"X-Operator-Token": settings.proxy_operator_token},
        )
        if response.is_error:
            raise RuntimeError(_response_error_message(response))
        return response.json()


async def get_operator_status() -> dict[str, Any]:
    settings = get_settings()
    async with httpx.AsyncClient(timeout=10.0) as client:
        response = await client.get(
            f"{settings.proxy_operator_url.rstrip('/')}/health",
            headers={"X-Operator-Token": settings.proxy_operator_token},
        )
        response.raise_for_status()
        return response.json()


async def ensure_wildcard_certificate() -> dict[str, Any]:
    settings = get_settings()
    async with httpx.AsyncClient(timeout=180.0) as client:
        response = await client.post(
            f"{settings.proxy_operator_url.rstrip('/')}/api/runtime/wildcard/ensure",
            headers={"X-Operator-Token": settings.proxy_operator_token},
        )
        if response.is_error:
            raise RuntimeError(_response_error_message(response))
        return response.json()
